sync3/caches/user.go

```diff
@@ -287,97 +258,32 @@ func (c *UserCache) OnRegistered(ctx context.Context, _ int64) error {
-		if len(urd.Timeline) > 0 && urd.LoadPos <= loadPos {
-			timeline := urd.Timeline
-			if len(timeline) > maxTimelineEvents {
-				timeline = timeline[len(timeline)-maxTimelineEvents:]
-			}
-			// ensure that if the user initially wants 1 event in this room then bumps up to 50 that
-			// we actually give them 50. This is more than just a length check as the room may not
-			// have 50 events, we can tell this based on the create event
-			createEventExists := false
-			if len(timeline) < maxTimelineEvents {
-				for _, ev := range timeline {
-					if gjson.GetBytes(ev, "type").Str == "m.room.create" && gjson.GetBytes(ev, "state_key").Str == "" {
-						createEventExists = true
-						break
-					}
-				}
-			}
-
-			// either we satisfied their request or we can't get any more events, either way that's good enough
-			if len(timeline) == maxTimelineEvents || createEventExists {
-				if !createEventExists && len(timeline) > 0 {
-					// fetch a prev batch token for the earliest event
-					_, ok := urd.PrevBatch()
-					if !ok {
-						eventID := gjson.ParseBytes(timeline[0]).Get("event_id").Str
-						prevBatch, err := c.store.EventsTable.SelectClosestPrevBatchByID(roomID, eventID)
-						if err != nil {
-							logger.Err(err).Str("room", roomID).Str("event_id", eventID).Msg("failed to get prev batch token for room")
-							internal.GetSentryHubFromContextOrDefault(ctx).CaptureException(err)
-						}
-						urd.SetPrevBatch(eventID, prevBatch)
-					}
-				}
-
-				// we already have data, use it
-				u := NewUserRoomData()
-				u.NotificationCount = urd.NotificationCount
-				u.HighlightCount = urd.HighlightCount
-				u.Timeline = timeline
-				u.PrevBatches = urd.PrevBatches
-				result[roomID] = u
-			} else {
-				// refetch from the db
-				lazyRoomIDs = append(lazyRoomIDs, roomID)
-			}
-		} else {
-			lazyRoomIDs = append(lazyRoomIDs, roomID)
-			// in case the room is left/invited, we may not add them to the result here so do it now,
-			// we'll clobber if we get a timeline
-			result[roomID] = urd
-		}
-	}
-	if len(lazyRoomIDs) == 0 {
-		return result
-	}
-	roomIDToEvents, roomIDToPrevBatch, err := c.store.LatestEventsInRooms(c.UserID, lazyRoomIDs, loadPos, maxTimelineEvents)
+	roomIDToEvents, roomIDToPrevBatch, err := c.store.LatestEventsInRooms(c.UserID, roomIDs, loadPos, maxTimelineEvents)
 	if err != nil {
-		logger.Err(err).Strs("rooms", lazyRoomIDs).Msg("failed to get LatestEventsInRooms")
+		logger.Err(err).Strs("rooms", roomIDs).Msg("failed to get LatestEventsInRooms")
 		internal.GetSentryHubFromContextOrDefault(ctx).CaptureException(err)
 		return nil
 	}
 	c.roomToDataMu.Lock()
-	for roomID, events := range roomIDToEvents {
-		urd, ok := c.roomToData[roomID]
+	for _, requestedRoomID := range roomIDs {
+		events := roomIDToEvents[requestedRoomID]
+		urd, ok := c.roomToData[requestedRoomID]
 		if !ok {
 			urd = NewUserRoomData()
 		}
-		oldLoadPos := urd.LoadPos
-		urd.Timeline = events
+		urd.RequestedTimeline = events
```

Comment:
Above we say that the new fields `RequestedTimeline` and `RequestedPrevBatch`

> are set by LazyLoadTimelines and are per-function call, and are not persisted in-memory.

But this `urd` comes from `c.roomToData` on a UserCache `c`, which sounds like it might be persistent. What's going on here?

Reply:
Ahh, `roomToData` holds UserRoomData values, so I am guessing that the local `urd` is a copy of (not a reference to) the cached struct.

Reply:
Bingo